Purchases/CodableExtensions/PurchaseOwnershipType+Extensions.swift

```diff
@@ -22,15 +22,26 @@ extension PurchaseOwnershipType: Decodable {
                                                   message: "Unable to extract an purchaseOwnershipTypeString")
         }
 
-        switch purchaseOwnershipTypeString {
-        case "PURCHASED":
-            self = .purchased
-        case "FAMILY_SHARED":
-            self = .familyShared
-        default:
+        if let type = Self.mapping[purchaseOwnershipTypeString] {
+            self = type
+        } else {
             Logger.error(Strings.codable.unexpectedValueError(type: PurchaseOwnershipType.self))
             self = .unknown
         }
     }
 
+    private static let mapping: [String: PurchaseOwnershipType] = PurchaseOwnershipType.allCases
+        .filter { $0.name != nil }
+        .dictionaryWithKeys { $0.name! }
```

Comment:
I don't love this, maybe I'll add a `Dictionary.mapKeys` or `compactMapKeys` so we can change it to:
```swift
PurchaseOwnershipType.allCases
  .dictionaryWithKeys { $0.name }
  .compactMapKeys { $0 }    
```

Reply:
I'm also torn here 🤷‍♂️ It _looks_ dangerous with the explicit unwrap but fine because that filter

How about a `reduce` like this?

```swift
private static let mapping: [String: Self] = Self.allCases
    .reduce(into: [:]) { (result, type) in
        if let name = type.name { result[name] = type }
    }
```

Reply:
Yeah I think it's not worth introducing a new operator just for this so `reduce` works fine. Thanks for the suggestion!

Reply:
the dict sounds safer to me. But if you did a dict, you wouldn't need this mapping in the first place, right? The dict _is_ the mapping? 
ie.: 

```swift
private static let ownershipTypesByName = [
    "PURCHASED": .purchased,
    "FAMILY_SHARED": .familyShared
]
```

and then

```swift
if let type = Self.ownershipTypesByName[purchaseOwnershipTypeString] {
    self = type
} else {
    Logger.error(Strings.codable.unexpectedValueError(type: PurchaseOwnershipType.self))
    self = .unknown
}
```

Reply:
oh, right, I missed that the point of this is to be able to switch on `PurchaseOwnershipType` to ensure that we don't miss a new case. nevermind

Reply:
Changed.